### PR TITLE
polish up, rebrand, add supabase auth

### PR DESCRIPTION
**Pull request by juliusmarminge**
## What kind of change does this PR introduce?

Bug fix, feature, docs update, ...

## What is the current behavior?

Please link any relevant issues here.

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

Add any other context or screenshots.

### CODE REVIEW

File: apps/expo/src/app/profile.tsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        cornerRadius={8}
+        onPress={async () => {
+          try {
+            const _credential = await AppleAuthentication.signInAsync({
+              requestedScopes: [
+                AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
+                AppleAuthentication.AppleAuthenticationScope.EMAIL,
+              ],
+            });
+            // signed in
```
Comment (juliusmarminge) [inline apps/expo/src/app/profile.tsx]: i guess we'll need to hook up the returned jwt here to supabase

https://docs.expo.dev/versions/latest/sdk/apple-authentication/

File: apps/expo/src/app/post/[id].tsx [hunk 2]
```diff
@@ -1,24 +1,22 @@
...
 
 import { api } from "../../utils/api";
 
-const Post: React.FC = () => {
+export default function PostPage() {
   const { id } = useSearchParams();
   if (!id || typeof id !== "string") throw new Error("unreachable");
   const { data } = api.post.byId.useQuery({ id });
 
-  if (!data) return <SplashScreen />;
```
Comment (juliusmarminge) [inline apps/expo/src/app/post/[id].tsx:11]: this loading state kinda sucked - maybe there should be a skeleton loader or something...

File: apps/expo/src/app/profile.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  );
+}
+
+function SignedOutView() {
+  const supabase = useSupabaseClient();
+
+  const signInWithGithub = () =>
+    supabase.auth
+      .signInWithOAuth({ provider: "github" })
+      .then((res) => console.log(res));
```
Comment (juliusmarminge) [inline apps/expo/src/app/profile.tsx]: fix this with new pkce flow

File: apps/expo/src/components/header.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { type ReactNode } from "react";
+import { Text, TouchableOpacity } from "react-native";
+import { Image } from "expo-image";
+import { useRouter } from "expo-router";
+import { Entypo, Ionicons } from "@expo/vector-icons";
+import { useUser } from "@supabase/auth-helpers-react";
+
+export function AuthAvatar() {
+  const user = useUser();
```
Comment (juliusmarminge) [inline apps/expo/src/components/header.tsx:9]: maybe query from db and return only what we need, maybe from the Profile table?

File: apps/expo/src/utils/api.tsx [hunk 5]
```diff
@@ -43,13 +44,19 @@ const getBaseUrl = () => {
...
+  const [queryClient] = useState(() => new QueryClient());
+  const [trpcClient] = useState(() =>
     api.createClient({
       transformer: superjson,
       links: [
         httpBatchLink({
           url: `${getBaseUrl()}/api/trpc`,
+          headers: () => {
+            const token = session?.access_token;
+            return { Authorization: token };
```
Comment (juliusmarminge) [inline apps/expo/src/utils/api.tsx]: verify this works as expected when sign in works properly

File: apps/nextjs/src/pages/index.tsx [hunk 6]
```diff
@@ -135,27 +152,39 @@ const Home: NextPage = () => {
...
-          {secretMessage && <span> - {secretMessage}</span>}
-        </p>
+      {!user && (
+        <Auth
+          redirectTo="http://localhost:3000/"
+          appearance={{ theme: ThemeSupa }}
+          supabaseClient={supabaseClient}
+          providers={["apple", "github"]}
+          socialLayout="vertical"
+          onlyThirdPartyProviders
```
Comment (juliusmarminge) [inline apps/nextjs/src/pages/index.tsx]: there is some weird flash with the positioning here when only using the oauth providers. maybe just remove this component and use normal buttons with callbacks?

File: packages/api/src/router/post.ts [hunk 7]
```diff
@@ -1,27 +1,72 @@
...
-    .mutation(({ ctx, input }) => {
-      return ctx.prisma.post.create({ data: input });
+    .mutation(async ({ ctx, input }) => {
+      function getNameFromUser() {
+        const meta = ctx.user.user_metadata;
+        if (typeof meta.name === "string") return meta.name;
+        if (typeof meta.full_name === "string") return meta.full_name;
+        if (typeof meta.user_name === "string") return meta.user_name;
+        return "[redacted]";
+      }
```
Comment (juliusmarminge) [inline packages/api/src/router/post.ts:37]: is this the best way 😨 - it's not terrible but it's not the prettiest :)

File: packages/api/src/router/post.ts [hunk 8]
```diff
@@ -1,27 +1,72 @@
...
+      return ctx.prisma.post.create({
+        data: {
+          author: {
+            connectOrCreate: {
+              where: { id: ctx.user.id },
+              create: {
+                id: ctx.user.id,
+                email: ctx.user.email,
+                name: getNameFromUser(),
+                image: ctx.user.user_metadata.avatar_url as string | undefined,
```
Comment (juliusmarminge) [inline packages/api/src/router/post.ts:48]: does every provider use the `avatar_url` field or should there be some `getImageFromUser` function like the name above to make sure this matches all possible fields?

File: packages/db/prisma/schema.prisma [hunk 9]
```diff
@@ -1,65 +1,301 @@
...
+  user   users   @relation(fields: [userId], references: [id], onDelete: Cascade)
+  posts  Post[]
+
+  @@schema("public")
+}
+
+// 
+// Supabase Generated
+
+model users {
```
Comment (juliusmarminge) [inline packages/db/prisma/schema.prisma:39]: would be good with a second pair of eyes here - is this the best way to do it?

I tried reading a bit and finding some other examples using supabase+prisma on how to sync the users. Maybe there is a prettier way? Do you need all the auth tables in this schema?

File: apps/expo/src/app/profile.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+              nonce: hashedNonce,
+            });
+
+            const idToken = credential.identityToken;
+
+            console.log(idToken);
+
+            if (!idToken) {
+              throw "ID Token not found";
+            }
```
Comment (dshukertjr) [inline apps/expo/src/app/profile.tsx]: Implemented Apple login, but couldn't verify that it works on my end.

**Comment by dshukertjr** [comment]
@juliusmarminge 
Thanks for the amazing work here! I just had some hick up, and it would be great if you could take a look and let me know if you know what maybe the cause!

I followed the steps on readme, and the third step is failing for me.
1. pnpm i
2. 1cp .env.example .env` and populated the credentials
3. pnpm db:push, which fails with the following error message

<details>
  <summary>error message in termial</summary>

```
• Packages in scope: @acme/api, @acme/db, @acme/eslint-config, @acme/expo, @acme/nextjs, @acme/tailwind-config
• Running db:push, db:generate in 6 packages
• Remote caching disabled
@acme/db:db:generate: cache bypass, force executing b82c987f0ec493df
@acme/db:db:push: cache bypass, force executing 289ccbbeceeb848e
@acme/db:db:generate: 
@acme/db:db:generate: > @acme/db@0.1.0 db:generate /Users/tylershukert/Desktop/create-t3-turbo/packages/db
@acme/db:db:generate: > pnpm with-env prisma generate
@acme/db:db:generate: 
@acme/db:db:push: 
@acme/db:db:push: > @acme/db@0.1.0 db:push /Users/tylershukert/Desktop/create-t3-turbo/packages/db
@acme/db:db:push: > pnpm with-env prisma db push --skip-generate
@acme/db:db:push: 
@acme/db:db:generate: 
@acme/db:db:generate: > @acme/db@0.1.0 with-env /Users/tylershukert/Desktop/create-t3-turbo/packages/db
@acme/db:db:generate: > dotenv -e ../../.env -- "prisma" "generate"
@acme/db:db:generate: 
@acme/db:db:push: 
@acme/db:db:push: > @acme/db@0.1.0 with-env /Users/tylershukert/Desktop/create-t3-turbo/packages/db
@acme/db:db:push: > dotenv -e ../../.env -- "prisma" "db" "push" "--skip-generate"
@acme/db:db:push: 
@acme/db:db:generate: Prisma schema loaded from prisma/schema.prisma
@acme/db:db:push: Prisma schema loaded from prisma/schema.prisma
@acme/db:db:push: Datasource "db": PostgreSQL database "postgres", schemas "auth, public" at "db.hvikvyyvkormucwjjohl.supabase.co:5432"
@acme/db:db:generate: 
@acme/db:db:generate: ✔ Generated Prisma Client (4.12.0 | library) to ./../../node_modules/@prisma/client in 125ms
@acme/db:db:generate: You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
@acme/db:db:generate: ```
@acme/db:db:generate: import { PrismaClient } from '@prisma/client'
@acme/db:db:generate: const prisma = new PrismaClient()
@acme/db:db:generate: ```
@acme/db:db:push: 
@acme/db:db:push: ⚠️  There might be data loss when applying the changes:
@acme/db:db:push: 
@acme/db:db:push:   • The primary key for the `schema_migrations` table will be changed. If it partially fails, the table could be left without primary key constraint.
@acme/db:db:push:   • You are about to alter the column `version` on the `schema_migrations` table, which contains 42 non-null values. The data in that column will be cast from `VarChar(255)` to `VarChar(14)`.
@acme/db:db:push:   • A unique constraint covering the columns `[version]` on the table `schema_migrations` will be added. If there are existing duplicate values, this will fail.
@acme/db:db:push: 
@acme/db:db:push: Error: Use the --accept-data-loss flag to ignore the data loss warnings like prisma db push --accept-data-loss
@acme/db:db:push:  ELIFECYCLE  Command failed with exit code 1.
@acme/db:db:push:  WARN   Local package.json exists, but node_modules missing, did you mean to install?
@acme/db:db:push:  ELIFECYCLE  Command failed with exit code 1.
@acme/db:db:push:  WARN   Local package.json exists, but node_modules missing, did you mean to install?
@acme/db:db:push: ERROR: command finished with error: command (/Users/tylershukert/Desktop/create-t3-turbo/packages/db) pnpm run db:push exited (1)
command (/Users/tylershukert/Desktop/create-t3-turbo/packages/db) pnpm run db:push exited (1)

 Tasks:    1 successful, 2 total
Cached:    0 cached, 2 total
  Time:    2.439s 
```

</details>

I also wanted to try out the Apple login, but kept getting `Unexpected error from Apple SignIn: AuthRetryableFetchError: Network request failed`. It looks like any network request going out from the iOS simulator is failing, but is there any setup process that I should take other than updating the `apps/expo/app.config.ts` file with the Supabase credential?

File: apps/nextjs/src/pages/index.tsx (hunk 6 above)
Comment (dshukertjr) [reply apps/nextjs/src/pages/index.tsx]: Personally, I think we don't need to nail every single details at this point, but if you want to replace it, that would be nice too!

File: apps/expo/src/components/header.tsx (hunk 4 above)
Comment (dshukertjr) [reply apps/expo/src/components/header.tsx:9]: I think this is totally fine.

File: packages/db/prisma/schema.prisma (hunk 9 above)
Comment (dshukertjr) [reply packages/db/prisma/schema.prisma:39]: I don't have much experience with Prisma, but these are auto generated from the existing tables within auth schema, correct? And if we don't include them here, all of these tables will be deleted, correct?

File: packages/api/src/router/post.ts (hunk 7 above)
Comment (dshukertjr) [reply packages/api/src/router/post.ts:37]: There is a way to use database triggers to copy over what we have in the metadata to public.profiles table
https://supabase.com/docs/guides/auth/managing-user-data#advanced-techniques

File: packages/db/prisma/schema.prisma (hunk 9 above)
Comment (juliusmarminge) [reply packages/db/prisma/schema.prisma:39]: Yea if you do a db push without these models they will be removed which we don't want

Comment (J0) [reply packages/db/prisma/schema.prisma:39]: Don't have much experience with Prisma either but it looks similar to our recommendation for our [Prisma integration guide](https://supabase.com/docs/guides/integrations/prisma) I'm not sure that there's a better way at the moment but let me check with the guide authors and get back to you

File: packages/api/src/router/post.ts (hunk 8 above)
Comment (J0) [reply packages/api/src/router/post.ts:48]: It's termed as `avatar_url` on most if not all of the major providers (e.g. Google, Github. etc). There may be a few providers which use a different name but it'd be hard to tell in advance - think it's fine to leave as is and if users are using a provider which doesn't use `avatar_url` as a field we can reconsider whether to adjust

**Comment by juliusmarminge** [comment]
> I also wanted to try out the Apple login, but kept getting Unexpected error from Apple SignIn: AuthRetryableFetchError: Network request failed. It looks like any network request going out from the iOS simulator is failing, but is there any setup process that I should take other than updating the apps/expo/app.config.ts file with the Supabase credential?

The Apple Auth only works on physical device i think: https://docs.expo.dev/versions/latest/sdk/apple-authentication#development-and-testing

As for the error you're getting - is it a new supabase project you're trying out or how is the setup there? It should just be as simple as creating a new project on Supabase then doing a db push i believe

EDIT: Oh it's the migrations table - I just use db:push so i accepted data loss on that one. I see though that your guide recommends using `prisma migrate` so maybe we should sync up the 2 readmes to be the same?

I'll do that tonight! For now, I think you can just db:push and accept the data loss on the migrations table

EDIT2: Looks like `prisma migrate` also complains. The issue is on this table:

```prisma
model schema_migrations {
  version String @unique(map: "schema_migrations_version_idx") @db.VarChar(14)

  @@schema("auth")
}
```

Not sure why the `prisma db pull` put that unique constraint on there if it's not actaulyl present in db???

**Review by dshukertjr — APPROVED** [review]
Thank you so much for the amazing work! This is awesome!

**Review by J0 — APPROVED** [review]
Lgtm

**Pull request merged**